ci: Upgrade most ci configs to focal

maflcko opened the pull request:
Generally developers compile with recent compilers, so bumping the ci configs to a recent OS should be uncontroversial. Older OSes (especially with compiler sanitizers) need workarounds that can be dropped by running on a more recent OS.

This pull changes the asan sanitizer and the experimental multiprocess build to use focal.
Also, it runs the no_wallet config on xenial to test against python 3.5, according to `doc/dependencies.md`.

Finally, all configs that mimic gitian (win and mac) will stay at bionic.

Commit:
doc: move doc to ci readme

Commit:
travis: Always run multiprocess build

Commit:
ci: Upgrade most ci configs to focal

hebasto commented on .travis.yml:
```diff
@@ -129,13 +115,12 @@ jobs:
 
     - stage: test
-      name: 'x86_64 Linux [GOAL: install]  [bionic]  [multiprocess]'
-      if: type != pull_request OR commit_message =~ /depends:|multiprocess:/ # Skip on non-depends, non-multiprocess PRs
```
Why this `if:` removed?

MarcoFalke replied:
I think it is not worth the risk of having the build go broken unnoticed as opposed to the cost of needing one more vm. The travis contract is expiring this month and we will either renew it or buy a contract at a different provider.

MarcoFalke commented:
For some reason this doesn't link:

```
/usr/bin/ld: qt/libbitcoinqt.a(libbitcoinqt_a-bitcoin.o): in function `GuiMain(int, char**)':

/home/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/qt/bitcoin.cpp:444: undefined reference to `qInitResources_bitcoin()'

/usr/bin/ld: /home/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/qt/bitcoin.cpp:445: undefined reference to `qInitResources_bitcoin_locale()'

clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

locally everything works fine, but on travis not: https://travis-ci.org/github/bitcoin/bitcoin/jobs/697962112#L4091

hebasto commented:
> For some reason this doesn't link:
> 
> ```
> /usr/bin/ld: qt/libbitcoinqt.a(libbitcoinqt_a-bitcoin.o): in function `GuiMain(int, char**)':
> 
> /home/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/qt/bitcoin.cpp:444: undefined reference to `qInitResources_bitcoin()'
> 
> /usr/bin/ld: /home/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/qt/bitcoin.cpp:445: undefined reference to `qInitResources_bitcoin_locale()'
> 
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```
> 
> locally everything works fine, but on travis not: https://travis-ci.org/github/bitcoin/bitcoin/jobs/697962112#L4091

Could you try to add `#include <QDir>` into `bitcoin.cpp`?

kiminuo commented:
(@hebasto Different PR but similar error. However, your suggestion did not help me: https://travis-ci.com/github/kiminuo/bitcoin/jobs/348750106 https://travis-ci.com/github/kiminuo/bitcoin/jobs/348750106#L4065)

Commit:
ci: Have one config run in xenial to test against python3.5

Also, bump the travis env to bionic. This shouldn't matter at all
because all ci configs run inside a docker, but it does seem to fix a
bug. See
https://github.com/bitcoin/bitcoin/pull/19267#issuecomment-643630309

hebasto commented:
@MarcoFalke 
It seem the root is here:
https://travis-ci.org/github/bitcoin/bitcoin/jobs/697962112#L3568

MarcoFalke commented:
Nice catch. Though,

* Why does this error not abort the build immediately?
* Why does it not reproduce locally?

hebasto commented:
Travis is happy with 17dc996ec5c7a53627396b0dbd6a2b61804b9ba4 :)

fanquake commented:
Concept ACK.

> Generally developers compile with recent compilers, so bumping the ci configs to a recent OS should be uncontroversial.

If we're going to do this I think we should make a decision in #18921. Otherwise the way we compile and test in our CI (as well as the way some developers are compiling locally) is starting to diverge from the way we actually build releases.

> For 0.21 this will be Ubuntu Bionic with python3.6.

Is the minimum supported Python being bumped to 3.6 for 0.21, or are we just not going to test Python 3.5 compat any more? You said [here](https://github.com/bitcoin/bitcoin/pull/18210#issuecomment-591458941):
> https://packages.ubuntu.com/xenial/python3 is still supported and not EOL until April next year, so we can't bump to python 3.6 yet.

MarcoFalke commented:
> starting to diverge from the way we actually build releases.

I don't mind bumping gitian to focal, but I think that discussion should happen elsewhere. I kept the three gitian-mimic configs at bionic (win64, mac, and the C++17 linux build). Though, sanitizers are not supported by gitian, so I think requiring them to use the older LTS and patch them with workarounds might not be the best way forward.

> Is the minimum supported Python being bumped to 3.6 for 0.21, or are we just not going to test Python 3.5 compat any more?

I am testing python 3.5 compat, but most of my testing has moved out of this repo. Though, I've switched the existing no_wallet config to xenial in the last commit to make it a bit more explicit that python3.5 compat is not changed by this pull.

practicalswift commented:
Concept ACK

MarcoFalke commented:
Force pushed to get rid of the already merged commit

laanwj commented:
Concept ACK. I think focal is a better reflection of the compiler and binutils most people are using than bionic, at this point.

Though I agree with @fanquake's point that we're still building releases using bionic, so we need to test that.

> However, there should be at least one config to test compile and run-time compatibility on the oldest supported OS. For 0.21 this will be Ubuntu Bionic with python3.6.

This is somewhat confusing to me—compile-time and run-time compatibility was never the same. Run-time compatibility (at least for pre-built releases) historically at least tends to back farther than we support building on. But I suppose backward-testing gitian releases isn't part of the CI so it doesn't matter here.

MarcoFalke commented:
`doc/dependencies.md` says it is python 3.5, so I went ahead and updated OP

MarcoFalke commented:
re-arranged commits to minimize diff

Sjors commented:
ACK fa05f44893d228f672f39436d0cb6b3376f81ac2, assuming Travis passes
(it hasn't shown up in the checks yet)

MarcoFalke commented:
>  (it hasn't shown up in the checks yet)

It shows for me:

![Screenshot_2020-06-19 ci Upgrade most ci configs to focal by MarcoFalke · Pull Request #19267 · bitcoin bitcoin](https://user-images.githubusercontent.com/6399679/85155235-82222a80-b226-11ea-9d89-9bc39bd8a955.png)

hebasto commented on .travis.yml:
```diff
@@ -126,13 +112,12 @@ jobs:
 
     - stage: test
-      name: 'x86_64 Linux  [GOAL: install]  [bionic]  [no wallet]'
+      name: 'x86_64 Linux  [GOAL: install]  [xenial]  [no wallet]'
```
Why `xenial`?

MarcoFalke replied:
Please refer to the commit message and pull request description for the motivation behind changes

hebasto replied:
I see.

hebasto reviewed (approved):
ACK fa05f44893d228f672f39436d0cb6b3376f81ac2

Sjors commented:
I show up for me now as well.
